### ISSUE TITLE
Make a side project that transforms the AST into a weakly typed AST

### ISSUE THREAD

**Issue by matozoid** [javaparser/javaparser#968]
The current AST is nice and type safe, but it is very hard to extend. We could create a weakly typed AST alternative that only has a Node class, but we would have to be able to switch the parser from generating one type of AST to another. So instead of having `new`s in the grammar, we could call a factory there, the default factory containing those `new`s. Now we can switch out the factory and create a different type of AST. Cool!

**Comment by ftomassetti** [comment]
Interesting!

**Comment by matozoid** [comment]
This won't work actually, since I can't vary the return types of the parser rules. A factory might return MyCustomExpression but the grammar still wants `com.github.javaparser.expr.Expression`.

Another way to do this is simply generate a visitor that turns the usual AST into another AST. Simple!

**Comment by matozoid** [comment]
So I'm thinking of the opposite of what we have now: have a single `Node` that contains a few default properties like a `TokenRange`, a `String type`, and a `Map<String, Object>` of properties. Those properties can be nodes, strings, lists, whatever. Useful for people who want to play with extending or otherwise mutilating Java. The project should have some support code too, maybe a simple functional interface for selecting and traversing stuff, and a printer.

**Comment by ftomassetti** [comment]
Makes sense. I think this project would be interesting and we should try to understand better the scenario in which it will be used and the requirements for that scenario.

In a more general way I think that as we gain more popularity we should have some sort of strategic sessions: we could gather and discuss where to focus our energies and try to understand what is most important for the project. The outcome should be a list of priorities and a sort of plan for the next 6 months.

I suggest this because there are so many things that look cool and I am not sure which ones we should pursue, given we have limited energies. It would be nice if someone started to sponsor JavaParser so that we could invest more time on it :)

**Comment by matozoid** [comment]
The Go people created a "Universal AST" defined here: https://doc.bblf.sh/uast/uast-specification.html

Using code generation to create a visitor that transforms our AST into that AST might be a good topic for the book.
